checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -368,52 +455,82 @@ private void analyzeOwnershipTransferAtAssignment(
+   * @param obligation the obligation
+   * @return the index of the current method parameter that is must-aliased to the given obligation,
+   *     if one exists; otherwise, return -1.
    */
-  private void addOwningToParamsIfDisposedAtAssignment(
-      Set<Obligation> obligations, Obligation rhsObligation, Node rhs) {
-    Set<ResourceAlias> rhsAliases = rhsObligation.resourceAliases;
-    if (rhsAliases.isEmpty()) {
-      return;
-    }
+  private int getIndexOfParam(Obligation obligation) {
+    Set<ResourceAlias> resourceAliases = obligation.resourceAliases;
     List<VariableElement> paramElts =
         CollectionsPlume.mapList(TreeUtils::elementFromDeclaration, methodTree.getParameters());
-    for (ResourceAlias rhsAlias : rhsAliases) {
-      Element rhsElt = rhsAlias.element;
-      int i = paramElts.indexOf(rhsElt);
-      if (i != -1) {
-        addOwningToParam(i + 1);
-        mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
-        break;
+    for (ResourceAlias resourceAlias : resourceAliases) {
+      int paramIndex = paramElts.indexOf(resourceAlias.element);
+      if (paramIndex != -1) {
+        return paramIndex + 1;
       }
     }
+
+    return -1;
+  }
+
+  /** Adds a {@link NotOwning} annotation to the return type of the current method. */
+  private void addNotOwningToReturnType() {
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    wpi.addMethodDeclarationAnnotation(methodElt, NOTOWNING);
+  }
+
+  /**
+   * Adds a {@link MustCallAlias} annotation to the formal parameter at the given index.
+   *
+   * @param index the index of a formal parameter of the current method (1-based)
+   */
+  private void addMustCallAliasToFormalParameter(int index) {
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    wpi.addMethodDeclarationAnnotation(methodElt, MUSTCALLALIAS);
+    wpi.addDeclarationAnnotationToFormalParameter(methodElt, index, MUSTCALLALIAS);
   }
 
   /**
    * Adds an {@link EnsuresCalledMethods} annotation to the current method for any owning field
    * whose must-call obligation is satisfied within the current method, i.e., the fields in {@link
    * #disposedFields}.
    */
-  private void addEnsuresCalledMethods() {
+  private void addEnsuresCalledMethodsForDisposedFields() {
     // The keys are the must-call method names, and the values are the set of fields on which those
-    // methods should be called. This map is used to create a single @EnsuresCalledMethods
-    // annotation for fields that share the same must-call obligation.
+    // methods should be called. This map is used to create a @EnsuresCalledMethods annotation for
+    // each set of fields that share the same must-call obligation.
     Map<String, Set<String>> methodToFields = new LinkedHashMap<>();
     for (VariableElement disposedField : disposedFields) {
       List<String> mustCallValues = resourceLeakAtf.getMustCallValues(disposedField);
```

Comment:
The value of `mustCallValues` was known when constructing `disposedFields`.  Would it make sense to record it it `dispasedFields` so it does not need to be recomputed here?

Reply:
I changed the `deposedFields` set to a map to keep the `must-call` value and save some computation here. Thanks for the suggestion.